File: app/src/main/java/com/franciscoolivero/android/weatherapp/view/WeatherActivity.java
package com.franciscoolivero.android.weatherapp.view;

import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;

import com.franciscoolivero.android.weatherapp.R;
import com.franciscoolivero.android.weatherapp.viewmodel.WeatherViewModel;
import com.google.android.material.appbar.AppBarLayout;
import com.google.android.material.tabs.TabLayout;
import com.google.android.material.tabs.TabLayoutMediator;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.lifecycle.ViewModelProviders;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import androidx.viewpager2.widget.ViewPager2;
import butterknife.BindView;
import butterknife.ButterKnife;

public class WeatherActivity extends AppCompatActivity {

    @BindView(R.id.coordinator_layout)
    CoordinatorLayout coordinatorLayout;

    @BindView(R.id.weather_tabs_layout)
    TabLayout tabLayout;

    @BindView(R.id.weather_toolbar)
    Toolbar toolbar;

    @BindView(R.id.weather_appbar)
    AppBarLayout appBarLayout;

    @BindView(R.id.weather_view_pager)
    ViewPager2 viewPager;

    @BindView(R.id.swipe_refresh_layout)
    SwipeRefreshLayout swipeRefreshLayout;

    private WeatherViewModel weatherViewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_weather);
        ButterKnife.bind(this);
        weatherViewModel = ViewModelProviders.of(this).get(WeatherViewModel.class);
        setupTabLayoutWithViewPager();
        setupSwipeRefreshListener();
        setSupportActionBar(toolbar);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.weather_options_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.more) {
            Intent settingsIntent = new Intent(this, SettingsActivity.class);
            startActivity(settingsIntent);
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void setupSwipeRefreshListener() {
        swipeRefreshLayout.setOnRefreshListener(() -> {
            weatherViewModel.fetchLocationAndWeatherData(true, this.getApplicationContext());
            swipeRefreshLayout.setRefreshing(false);
        });
    }

    private void setupTabLayoutWithViewPager() {
        viewPager.setUserInputEnabled(false);
        TabsPagerAdapter tabsPagerAdapter = new TabsPagerAdapter(this);
        viewPager.setAdapter(tabsPagerAdapter);
        new TabLayoutMediator(tabLayout, viewPager,
                (tab, position) -> {
                    if (position == 0) {
                        tab.setText(getResources().getString(R.string.today_tab));
                    } else {
                        tab.setText(getResources().getString(R.string.daily_tab));
                    }
                }).
                attach();
    }
}

File: app/src/main/java/com/franciscoolivero/android/weatherapp/view/WeatherActivity.kt
package com.franciscoolivero.android.weatherapp.view

import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.lifecycle.ViewModelProviders
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import androidx.viewpager2.widget.ViewPager2
import butterknife.BindView
import butterknife.ButterKnife
import com.franciscoolivero.android.weatherapp.R
import com.franciscoolivero.android.weatherapp.viewmodel.WeatherViewModel
import com.google.android.material.appbar.AppBarLayout
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayoutMediator
import com.google.android.material.tabs.TabLayoutMediator.TabConfigurationStrategy

class WeatherActivity : AppCompatActivity() {
    @JvmField
    @BindView(R.id.coordinator_layout)
    var coordinatorLayout: CoordinatorLayout? = null

    @JvmField
    @BindView(R.id.weather_tabs_layout)
    var tabLayout: TabLayout? = null

    @JvmField
    @BindView(R.id.weather_toolbar)
    var toolbar: Toolbar? = null

    @JvmField
    @BindView(R.id.weather_appbar)
    var appBarLayout: AppBarLayout? = null

    @JvmField
    @BindView(R.id.weather_view_pager)
    var viewPager: ViewPager2? = null

    @JvmField
    @BindView(R.id.swipe_refresh_layout)
    var swipeRefreshLayout: SwipeRefreshLayout? = null
    private var weatherViewModel: WeatherViewModel? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_weather)
        ButterKnife.bind(this)
        weatherViewModel = ViewModelProviders.of(this).get(WeatherViewModel::class.java)
        setupTabLayoutWithViewPager()
        setupSwipeRefreshListener()
        setSupportActionBar(toolbar)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.weather_options_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == R.id.more) {
            val settingsIntent = Intent(this, SettingsActivity::class.java)
            startActivity(settingsIntent)
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private fun setupSwipeRefreshListener() {
        swipeRefreshLayout!!.setOnRefreshListener {
            weatherViewModel!!.fetchLocationAndWeatherData(true, this.applicationContext)
            swipeRefreshLayout!!.isRefreshing = false
        }
    }

    private fun setupTabLayoutWithViewPager() {
        viewPager!!.isUserInputEnabled = false
        val tabsPagerAdapter = TabsPagerAdapter(this)
        viewPager!!.adapter = tabsPagerAdapter
        TabLayoutMediator(tabLayout!!, viewPager!!,
                TabConfigurationStrategy { tab: TabLayout.Tab, position: Int ->
                    if (position == 0) {
                        tab.text = resources.getString(R.string.today_tab)
                    } else {
                        tab.text = resources.getString(R.string.daily_tab)
                    }
                }).attach()
    }
}